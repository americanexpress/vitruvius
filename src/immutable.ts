/*
 * Copyright (c) 2017 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */

import { combineReducers } from 'redux-immutable';
import { Map } from 'immutable';
import collectBuiltState, { VitruviusImmutableReducers } from './collectBuiltState';

// internal type for use before assigning buildInitial state
type VitruviusImmutableOptionalCombinedReducer<Locals> = ReturnType<typeof combineReducers> & {
  buildInitialState?: (_locals: Locals) => Map<string, unknown>;
};

export type VitruviusImmutableCombinedReducer<Locals> = ReturnType<typeof combineReducers> & {
  buildInitialState: (_locals: Locals) => Map<string, unknown>;
};

export default function vitruviusImmutable<Locals>(
  reducers: VitruviusImmutableReducers<Locals>,
  getDefaultState = Map<string, unknown>
): VitruviusImmutableCombinedReducer<Locals> {
  const combined: VitruviusImmutableOptionalCombinedReducer<Locals> = combineReducers(
    reducers,
    getDefaultState
  ) as VitruviusImmutableOptionalCombinedReducer<Locals>;

  combined.buildInitialState = function buildInitialState(locals): Map<string, unknown> {
    return collectBuiltState<Locals, VitruviusImmutableReducers<Locals>>({
      reducers,
      locals,
      defaultState: getDefaultState(),
    });
  };

  return combined as VitruviusImmutableCombinedReducer<Locals>;
}
