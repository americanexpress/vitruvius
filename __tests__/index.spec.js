import { Map } from 'immutable';
import { createStore } from 'redux';
import vitruvius from '../src';
import {
  stuffReducer,
  thingsReducer,
  otherReducer,
  staticReducer,
} from './fixtures';

describe('vitruvius', () => {
  const locals = { data: 'test' };

  it('should handle all reducers having buildInitialState method', () => {
    const reducer = vitruvius({
      stuff: stuffReducer,
      things: thingsReducer,
      other: otherReducer,
    });
    const actual = reducer.buildInitialState(locals);
    const expected = {
      stuff: new Map({ foo: locals.data }),
      things: new Map({ bar: locals.data }),
      other: new Map({ baz: locals.data }),
    };
    expect(actual).toEqual(expected);
  });

  it('should handle no reducers having buildInitialState method', () => {
    const reducer = vitruvius({
      static: staticReducer,
    });
    const actual = reducer.buildInitialState(locals);
    expect(actual).toEqual({});
  });

  it('should handle a mix of reducers having and not having buildInitialState method', () => {
    const reducer = vitruvius({
      stuff: stuffReducer,
      things: thingsReducer,
      other: otherReducer,
      static: staticReducer,
    });
    const actual = reducer.buildInitialState(locals);
    const expected = {
      stuff: new Map({ foo: locals.data }),
      things: new Map({ bar: locals.data }),
      other: new Map({ baz: locals.data }),
    };
    expect(actual).toEqual(expected);
  });

  it('should build the expected state for a flat tree of reducers', () => {
    const reducer = vitruvius({
      stuff: stuffReducer,
      things: thingsReducer,
      other: otherReducer,
      static: staticReducer,
    });
    const actual = reducer.buildInitialState(locals);
    const expected = {
      stuff: new Map({ foo: locals.data }),
      things: new Map({ bar: locals.data }),
      other: new Map({ baz: locals.data }),
    };
    expect(actual).toEqual(expected);
  });

  it('should build the expected state for a nested tree of reducers', () => {
    const reducer = vitruvius({
      stuff: stuffReducer,
      nested: vitruvius({
        things: thingsReducer,
        other: otherReducer,
      }),
    });
    const actual = reducer.buildInitialState(locals);
    const expected = {
      stuff: new Map({ foo: locals.data }),
      nested: {
        things: new Map({ bar: locals.data }),
        other: new Map({ baz: locals.data }),
      },
    };
    expect(actual).toEqual(expected);
  });

  it('should return an initialState that is acceptable to redux\'s createStore', () => {
    const reducer = vitruvius({
      stuff: stuffReducer,
      things: thingsReducer,
      other: otherReducer,
      static: staticReducer,
    });
    const store = createStore(reducer, reducer.buildInitialState(locals));
    const actual = store.getState();
    const expected = {
      stuff: new Map({ foo: locals.data }),
      things: new Map({ bar: locals.data }),
      other: new Map({ baz: locals.data }),
      static: new Map({ static: true }),
    };
    expect(actual).toEqual(expected);
  });
});
