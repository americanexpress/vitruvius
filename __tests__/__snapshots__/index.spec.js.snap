// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vitruvius should build the expected state for a flat tree of reducers 1`] = `
{
  "other": Immutable.Map {
    "baz": "test",
  },
  "stuff": Immutable.Map {
    "foo": "test",
  },
  "things": Immutable.Map {
    "bar": "test",
  },
}
`;

exports[`vitruvius should build the expected state for a nested tree of reducers 1`] = `
{
  "nested": {
    "other": Immutable.Map {
      "baz": "test",
    },
    "things": Immutable.Map {
      "bar": "test",
    },
  },
  "stuff": Immutable.Map {
    "foo": "test",
  },
}
`;

exports[`vitruvius should handle a mix of reducers having and not having buildInitialState method 1`] = `
{
  "other": Immutable.Map {
    "baz": "test",
  },
  "stuff": Immutable.Map {
    "foo": "test",
  },
  "things": Immutable.Map {
    "bar": "test",
  },
}
`;

exports[`vitruvius should handle all reducers having buildInitialState method 1`] = `
{
  "other": Immutable.Map {
    "baz": "test",
  },
  "stuff": Immutable.Map {
    "foo": "test",
  },
  "things": Immutable.Map {
    "bar": "test",
  },
}
`;

exports[`vitruvius should handle no reducers having buildInitialState method 1`] = `{}`;

exports[`vitruvius should return an initialState that is acceptable to redux toolkit's configureStore 1`] = `
{
  "other": Immutable.Map {
    "baz": "foo",
  },
  "static": Immutable.Map {
    "static": true,
  },
  "stuff": Immutable.Map {
    "foo": "bar",
  },
  "things": Immutable.Map {
    "bar": "baz",
  },
}
`;

exports[`vitruvius should return an initialState that is acceptable to redux's createStore 1`] = `
{
  "other": Immutable.Map {
    "baz": "test",
  },
  "static": Immutable.Map {
    "static": true,
  },
  "stuff": Immutable.Map {
    "foo": "test",
  },
  "things": Immutable.Map {
    "bar": "test",
  },
}
`;
